Fix gap fill ruining some curves

kasbah opened the pull request:
working on #11

kasbah committed:
Add test for gap filling closest points

kasbah committed:
Add gap filling function that finds closest points

kasbah commented:
@mcous could you clue me in as to [why the depth-first traversal is needed](https://github.com/mcous/gerber-plotter/blob/9e9a81d38db8641bdb07a59fcd2d2a740d1fd910/test/path-graph_test.js#L56)?

mcous commented:
> could you clue me in as to why the depth-first traversal is needed?

@kasbah for the outline mask to work in the board, the outline path needs to be organized into loops. A depth-first traversal of the nodes ensures that it traverses full loops first

kasbah committed:
Improve gap filling function

kasbah committed:
Make path graph do a depth first traversal again

kasbah commented:
Thanks for the info. I guess an issue would only surface in some cases? I made it do a depth-first traversal as before again but I ran the pcb-stackup integration tests before fixing it and couldn't see any problems. I used an `outlineGapFill` of 1.0mm and it works well, especially with these two cases that I couldn't make work before. Doing the depth-first traversal the results look the same to me. 

![](http://cloud.monostable.co.uk/usbvil_top.svg) 
![](http://cloud.monostable.co.uk/chaos-looper_top.svg)

kasbah committed:
Remove unused fillGaps argument in pointsEqual

mcous commented on lib/path-graph.js:
```diff
@@ -16,16 +16,25 @@ var find = function (collection, condition) {
+    var d = distance(position, point.position)
+    if (d < fillGaps && d < prev.distance) {
+      return {point: point, distance: d}
+    } else {
```
Nitpick, but this else isn't necessary due to the early return

mcous reviewed (approved):
Fine by me if integration tests check out + masking seems improved!

kasbah committed:
Remove superfluous else

kasbah commented:
Cool, thanks for the review. So, just curios: keep the depth first traversal, it's _definitely_ needed? Or, keep it but just _in case_ it's actually needed?

mcous commented:
It looks like you've changed some of the graph construction logic around, so I don't much care how its traversed. What is vital is that whatever path is returned places segments that make up complete loops sequentially. Without that, `maskWithOutline` could fail depending on how the outline gerber is constructed

kasbah commented:
I feel like I need to find a proper test case for it before I feel comfortable removing it. I guess it might help that now the unconnected points join up at their nearest neighbor rather than bunching all points onto the first added point? I am not really sure about any of this though, the mutating structures in this were kicking my butt for quite a while.

kasbah committed:
Remove depth first traversal and test for it

kasbah commented:
So I managed to remove the depth-first traversal on the master branch and saw how that would mess up the integration test. Then I experimented with how much I could remove from this pull-request and basically I can just remove the "consolidation" `forEach` I added to make the test pass. I removed that and everything looks fine on the integration test. 

I am still not 100% happy as I don't really understand what it's doing now. It adds itself as an edge to the point in the `points` array so if another segment picks that point it will be marked as connecting? So when it traverses it it does those in order. I still don't really know if that makes it do "complete loops" in all cases though.

kasbah committed:
Revert "Remove depth first traversal and test for it"

This reverts commit b762a44574c4e558af8f1ef55927e776c944b7a6.

kasbah commented:
So take this example of a square drawn out of order:

```js
p.add({type: 'line', start: [0, 0], end: [3, 0]})
p.add({type: 'line', start: [3, 0], end: [3, 1]})
p.add({type: 'line', start: [3, 2], end: [3, 3]})
p.add({type: 'line', start: [3, 3], end: [0, 3]})
p.add({type: 'line', start: [3, 1], end: [3, 2]})
p.add({type: 'line', start: [0, 3], end: [0, 0]})
```

Which is drawn in this order: 

![image](https://user-images.githubusercontent.com/206854/38615509-708bb3b2-3d80-11e8-8db6-18fb0c0e5b24.png)

It doesn't re-order with the depth-first removed. Is it an issue? If I put the depth-first traversal back it re-orders it: 

![image](https://user-images.githubusercontent.com/206854/38615654-f235b5de-3d80-11e8-88ae-f8c5d27e99c3.png)

mcous commented:
Here is a simplified example of how the board outline is applied, and why I think that depth-first traversal is needed. Compare how the `clipPath` is applied to the rectangle when the path is out of order vs. how it is applied when the path is in order: https://codepen.io/mcous/pen/rdRgMB?editors=1000#0

([MDN's tutorial on `<path>` if you need it](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Paths))

kasbah commented:
Thanks for that, makes sense, I think I'll err on the side of caution with this and keep it like it was. The only downside to doing it is a unknown performance gain anyway.

kasbah committed:
Add comments about depth first traversal
